### PR TITLE
bpo-23689: re module, fix memory leak when a match is terminated by a signal or memory allocation failure

### PR DESCRIPTION
**Pull request by ghost**
@serhiy-storchaka 

This time, I checked several rounds carefully, it should be in good state.

Tested with `VERBOSE`/`VVERBOSE` macros defined, it builds and runs well. 

https://bugs.python.org/issue23689

### CODE REVIEW

File: Modules/_sre.c [hunk 1]
```diff
@@ -1837,17 +1849,25 @@ _validate_inner(SRE_CODE *code, SRE_CODE *end, Py_ssize_t groups)
...
                 if (min > max)
                     FAIL;
                 if (max > SRE_MAXREPEAT)
                     FAIL;
-                if (!_validate_inner(code, code+skip-3, groups))
+                if (op1 == SRE_OP_REPEAT) {
+                    GET_ARG; repeat_index = arg;
+                    if (repeat_index >= (size_t)self->repeat_count)
+                        FAIL;
+                    _fields = 4;
```
Comment (serhiy-storchaka) [inline Modules/_sre.c]: Maybe just `skip -= 4`?

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

I have only one question: how to prove that we need only one SRE_REPEAT structure per the REPEAT code?

File: Lib/test/test_re.py [hunk 2]
```diff
@@ -2338,8 +2338,8 @@ def test_possesive_repeat(self):
 ''')
 
     def test_repeat_index(self):
-        self.assertEqual(get_debug_out(r'(?:ab)*(?:cd)*'), '''\
-MAX_REPEAT 0 MAXREPEAT
+        self.assertEqual(get_debug_out(r'(?:ab)*?(?:cd)*'), '''\
+MIN_REPEAT 0 MAXREPEAT
```
Comment (serhiy-storchaka) [inline Lib/test/test_re.py:2342]: You just read my mind! I was going to propose such a change, but I thought that I was already bothering you too much.

Comment (ghost) [reply Lib/test/test_re.py:2342]: I just thought this after posting this PR.

> I thought that I was already bothering you too much.

As an inactive contributor, this is not a matter.
I'm not practised, so need continuously improve the patch to get to a good state.
When I think it's good, I can always find its shortcomings afterwards.  

> I have only one question: how to prove that we need only one SRE_REPEAT structure per the REPEAT code?

I have to think about how to answer your question.

**Comment by animalize** [comment]
> I have only one question: how to prove that we need only one SRE_REPEAT structure per the REPEAT code?

At any time, an `SRE_OP_REPEAT` only has one in the stack.
- When executing this OP, it's pushed in the stack. Although `SRE_OP_MAX_UNTIL` / `SRE_OP_MIN_UNTIL` may generate many backtracking points in the stack (above `SRE_OP_REPEAT` in the stack).
- When backtracking, it's poped from the stack.

This wouldn't work if re engine could memorize some backtracking state to optimize performance, but then the code of re module would be much more complicated.

**Pull request merged**